src/number_theory/modular_forms/slash_actions.lean

```diff
@@ -26,38 +25,38 @@ local notation `GL(` n `, ` R `)`⁺ := matrix.GL_pos (fin n) R
 local notation `SL(` n `, ` R `)` := matrix.special_linear_group (fin n) R
 
 /--A general version of the slash action of the space of modular forms.-/
-class slash_action (β : Type*) (G : Type*) (α : Type*) (γ : Type*) [group G] [ring α]
-  [has_smul γ α] :=
-(map : β → G → α → α)
+class slash_action (β G α γ : Type*) [group G] [has_zero α ]
+[has_one α ] [has_smul γ α ] [has_add α ] :=
+(map : β → G → α → α )
 (mul_zero : ∀ (k : β) (g : G), map k g 0 = 0)
 (one_mul : ∀ (k : β) (a : α) , map k 1 a = a)
 (right_action : ∀ (k : β) (g h : G) (a : α), map k h (map k g a) = map k (g * h) a )
 (smul_action : ∀ (k : β) (g : G) (a : α) (z : γ), map k g (z • a) = z • (map k g a))
 (add_action : ∀ (k : β) (g : G) (a b : α), map k g (a + b) = map k g a + map k g b)
 
 /--Slash_action induced by a monoid homomorphism.-/
-def monoid_hom_slash_action { β : Type*} {G : Type*} {H : Type*} {α : Type*} {γ : Type*}
-  [group G] [ring α] [has_smul γ α] [group H] [slash_action β G α γ] (h : H →* G) :
-  slash_action β H α γ:=
-{ map := (λ k g a, slash_action.map γ k (h(g)) a),
+def monoid_hom_slash_action {β G H α γ : Type*} [group G] [has_zero α]
+  [has_one α] [has_smul γ α ] [has_add α ] [group H] [slash_action β G α γ]
+  (h : H →* G) : slash_action β H α γ :=
+{ map := λ k g , slash_action.map γ k (h g) ,
   mul_zero := by {intros k g, apply slash_action.mul_zero k (h g), },
   one_mul := by {intros k a, simp only [map_one], apply slash_action.one_mul,},
   right_action := by {simp only [map_mul], intros k g gg a, apply slash_action.right_action,},
   smul_action := by {intros k g a z, apply slash_action.smul_action, },
```

Comment:
These are all a bit wordy, for example this one can be
```suggestion
  smul_action := λ _ _, slash_action.smul_action _ _,
```
Can you edit the others?  Same below.